osu.Game/Database/LegacyModelExporter.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using osu.Framework.Graphics;
+using osu.Framework.Logging;
+using osu.Framework.Platform;
+using osu.Game.Extensions;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
+using Realms;
+using SharpCompress.Archives.Zip;
+
+namespace osu.Game.Database
+{
+    /// <summary>
+    /// A class which handles exporting legacy user data of a single type from osu-stable.
+    /// </summary>
+    public abstract class LegacyModelExporter<TModel> : Component
+        where TModel : RealmObject, IHasNamedFiles, IHasGuidPrimaryKey
+    {
+        /// <summary>
+        /// The file extension for exports (including the leading '.').
+        /// </summary>
+        protected abstract string FileExtension { get; }
+
+        protected Storage UserFileStorage;
+        protected Storage ExportStorage;
+
+        protected RealmAccess RealmAccess;
+
+        private readonly ProgressNotification notification;
+
+        private string filename = "";
+
+        private bool canCancel = true;
+
+        protected LegacyModelExporter(Storage storage, RealmAccess realm, INotificationOverlay? notifications = null)
+        {
+            ExportStorage = storage.GetStorageForDirectory(@"exports");
+            UserFileStorage = storage.GetStorageForDirectory(@"files");
+            RealmAccess = realm;
+
+            notification = new ProgressNotification
+            {
+                State = ProgressNotificationState.Active,
+                Text = "Exporting...",
+                CompletionText = "Export completed"
+            };
+            notification.CancelRequested += () => canCancel;
+
+            notifications?.Post(notification);
+        }
+
+        /// <summary>
+        /// Export the model to default folder.
+        /// </summary>
+        /// <param name="item">The model should export.</param>
+        /// <returns></returns>
+        public async Task ExportAsync(RealmObject item)
+        {
+            if (item is TModel model)
+            {
+                filename = $"{model.GetDisplayString().GetValidFilename()}{FileExtension}";
+
+                using (var stream = ExportStorage.CreateFileSafely(filename))
+                {
+                    await ExportToStreamAsync(model, stream);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Export te model corresponding to uuid to given stream.
+        /// </summary>
+        /// <param name="uuid">The medel which have <see cref="IHasGuidPrimaryKey"/>.</param>
```

Comment:
The model to export?

Reply:
FWIW I basically ignored all xmldocs when reviewing this and will probably rewrite them in a final pass. Basically only wanted to ask about structure for now.